### PR TITLE
H23 view wallet

### LINKED ISSUE #130: H.2.3 View Wallet

### LINKED ISSUE #132: H.3.1 View Transaction History

### LINKED ISSUE #177: B.5.15 View Transaction History as Admin

### PR DESCRIPTION
**Pull request by ianleck**
## Description:
- View Wallet
- Added billings to checkouts
- Seeded wallet (Lesser evil in the end. Had other issues when I tried to just create during initialisation)
- View all billings

Flow for checkout: 
When check out,
need to:
1. Make billing from student to admin, pend payment
2. Update billing with paymentId after link has been received from paypal
3. Return paypal url to frontend

When course billing is captured, 
Need to:
1. Create courseContract for user
2. Update billing with payerId
3. Calculate Revenue + Earnings
4. Create billings from admin to sensei for the relevant amounts.
5. Update admin wallet
6. Return success message to frontend

## Checklist:
- [x] Merged latest develop
- [x] PR title makes sense

### CODE REVIEW

File: src/constants/enum.ts [hunk 1]
```diff
@@ -12,12 +12,27 @@ export enum ADMIN_VERIFIED_ENUM {
...
+export enum BILLING_ACTION {
+  AUTHORIZE = 'AUTHORIZE',
+  CAPTURE = 'CAPTURE',
+}
+
+export enum BILLING_STATUS {
   SUCCESS = 'SUCCESS',
   FAILED = 'FAILED',
+  PENDING_PAYMENT = 'PENDING_PAYMENT',
+  PENDING_90_DAYS = 'PENDING_90_DAYS',
```
Comment (kengthong) [inline src/constants/enum.ts:24]: Supposed to be 120 days right?

File: src/controllers/paypal.controller.ts [hunk 2]
```diff
@@ -27,6 +30,9 @@ export class PaypalController {
           });
           //if redirect url present, redirect user
           if (links.hasOwnProperty('approval_url')) {
+            const paymentId = links['self'].href.split('/')[6];
+            await WalletService.updatePaymentId(billings, paymentId);
+
```
Comment (kengthong) [inline src/controllers/paypal.controller.ts:35]: Line 37: got to add a response and not just a console log. Else nodejs server will hang

File: src/controllers/paypal.controller.ts [hunk 3]
```diff
@@ -59,7 +66,11 @@ export class PaypalController {
             console.error(error);
           } else {
             if (payment.state == 'approved') {
-              await PaypalService.captureOrder(paymentId, payerId.payer_id);
+              await PaypalService.captureOrder(
```
Comment (kengthong) [inline src/controllers/paypal.controller.ts:73]: Line 66: return response error too.

File: src/migrations/20210308062912-billing.js [hunk 4]
```diff
@@ -28,12 +39,37 @@ module.exports = {
...
+          type: Sequelize.DATE,
+        },
+        withdrawnDate: {
+          type: Sequelize.DATE,
+        },
+        withdrawalApplication: {
+          allowNull: false,
+          type: Sequelize.BOOLEAN,
+          defaultValue: false,
+        },
```
Comment (kengthong) [inline src/migrations/20210308062912-billing.js]: Could you leave some comments to elaborate what these fields are for and when will they be used?

File: src/routes/wallet.route.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+  Utility.asyncHandler(WalletController.viewWallet)
+);
+
+router.get(
+  '/:walletId/:billingId',
+  passport.authenticate('isAuthenticated', { session: false }),
+  requireFinance,
+  schemaValidator.params(wallet.billingIdP),
+  Utility.asyncHandler(WalletController.viewBilling)
+);
```
Comment (kengthong) [inline src/routes/wallet.route.ts]: For the permission check, can a sensei view his own wallet or will that be added in the future?

**Review by kengthong — CHANGES_REQUESTED** [review]

File: src/routes/wallet.route.ts (hunk 5 above)
Comment (ianleck) [reply src/routes/wallet.route.ts]: This permission check only blocks you if you are admin and not finance. Else anyone else can view

Comment (ianleck) [reply src/routes/wallet.route.ts]: Students will also get transaction history through view wallet, just that they won't be able to see the wallet on the frontend as student

**Review by kengthong — DISMISSED** [review]

**Review by kengthong — APPROVED** [review]

**Pull request merged**